packages/circuits-sdk/src/implementations/CircomWrapper.ts

```diff
@@ -1,29 +1,108 @@
-import { ObjectUtils } from "@snickerdoodlelabs/common-utils";
-import { CircuitError, JSONString, ZKProof } from "@snickerdoodlelabs/objects";
+import { ObjectUtils, IAxiosAjaxUtils } from "@snickerdoodlelabs/common-utils";
+import {
+  CircuitError,
+  JSONString,
+  ZKProof,
+  IpfsCID,
+  AjaxError,
+} from "@snickerdoodlelabs/objects";
 import { injectable, unmanaged } from "inversify";
-import { ResultAsync } from "neverthrow";
+import { ResultAsync, errAsync, okAsync } from "neverthrow";
 import { CircuitSignals, Groth16Proof, PublicSignals, groth16 } from "snarkjs";
+import { urlJoin } from "url-join-ts";
 
+import { ICircuitsSDKConfigProvider } from "@circuits-sdk/ICircuitsSDKConfigProvider.js";
+import { ICircomWrapper } from "@circuits-sdk/interfaces/ICircomWrapper.js";
 @injectable()
 export abstract class CircomWrapper<
   TInput extends CircuitSignals,
   TVerifyInput extends PublicSignals,
-> {
+> implements ICircomWrapper
+{
+  private _wasmResult?: ResultAsync<Uint8Array, AjaxError | CircuitError>;
+  private _zkeyResult?: ResultAsync<Uint8Array, AjaxError | CircuitError>;
   public constructor(
-    @unmanaged() protected wasm: Uint8Array,
-    @unmanaged() protected zkey: Uint8Array,
+    @unmanaged() protected ajaxUtils: IAxiosAjaxUtils,
+    @unmanaged() protected circuitsSDKConfig: ICircuitsSDKConfigProvider,
     @unmanaged() protected vkey: Record<string, unknown>,
+    @unmanaged() protected wasmKey: IpfsCID,
+    @unmanaged() protected zkeyKey: IpfsCID,
   ) {}
 
+  public preFetch(): ResultAsync<undefined, never> {
+    return this.circuitsSDKConfig.getConfig().andThen((config) => {
```

Comment:
One, you can just use .map() here to avoid the okAsync(), but 2, you should be returning the results from the fetchResource(). That way, you can chain preFetch() properly if you want- you just don't have to. If you return the result of this.fetchResource(), you'll need the .andThen() BTW, don't let me confuse you.